Source/WebGPU/WebGPU/BindGroup.mm

```diff
@@ -108,6 +108,8 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
                     return BindGroup::createInvalid(*this);
 
                 [argumentEncoder[stage] setBuffer:buffer offset:entry.offset atIndex:index++];
+                ASSERT(limits().maxBufferSize < UINT32_MAX);
+                *(uint32_t*)[argumentEncoder[stage] constantDataAtIndex:index++] = static_cast<uint32_t>(entry.size);
```

Comment:
Maybe we should do the signed one too while we're here? Shouldn't we be checking for overflow? We have `Checked` types for making that easy.

I think we should also be careful about units - any computation the shader has to perform will be executed millions of times per draw call, whereas any computation we can hoist to here will be executed once per bind group. If the value here is in "bytes" units, the shader will have to convert that to "indices" to bounds check against; if we can do that unit conversion here, that would probably be better in the end. (of course, that requires getting more info from the compiler about the size of the unsized array's inner type...)

Also, it sure would be good if there were a single shared function that would lower the BGL into a description of where everything in the Metal argument buffer will end up, rather than doing it like this via convention.

Reply:
oh, maybe we can't use the correct unit here because this code has to be able to run before any shaders are ever compiled

okay :|

Reply:
@litherum I don't know if we need to check for overflow here since there is no Metal device that supports anywhere close to a 2^32 - 1 buffer size. I think the current max Metal buffer size is 256 or 512MB.

If we want to perform the calculation once, I would suggest we pass the buffer sizes outside of the argument buffer via `-[MTLRenderCommandEncoder setVertexBytes:length:atIndex:]` for instance

Reply:
based on some offline discussion I will implement the buffer lengths in a separate buffer